pool/pool.go

```diff
@@ -541,25 +543,154 @@ func (p *pool) checkSessionTokenErr(err error, address string) bool {
 	return false
 }
 
-func (p *pool) PutObject(ctx context.Context, params *client.PutObjectParams, opts ...CallOption) (*oid.ID, error) {
+type callContext struct {
+	// context for RPC
+	ctxBase context.Context
+
+	client Client
+
+	// client endpoint
+	endpoint string
+
+	// request signer
+	key *ecdsa.PrivateKey
+
+	// flag to open default session if session token is missing
+	sessionDefault bool
+	sessionToken   *session.Token
+}
+
+func (p *pool) prepareCallContext(ctx *callContext, cfg *callConfig) error {
+	cp, err := p.connection()
+	if err != nil {
+		return err
+	}
+
+	ctx.endpoint = cp.address
+	ctx.client = cp.client
+
+	ctx.key = cfg.key
+	if ctx.key == nil {
+		ctx.key = p.key
+	}
+
+	ctx.sessionDefault = cfg.useDefaultSession
+	ctx.sessionToken = cfg.stoken
+
+	if ctx.sessionToken == nil && ctx.sessionDefault {
+		cacheKey := formCacheKey(ctx.endpoint, ctx.key)
+
+		ctx.sessionToken = p.cache.Get(cacheKey)
+		if ctx.sessionToken == nil {
+			var cliPrm client.CreateSessionPrm
+
+			cliPrm.SetExp(math.MaxUint32)
+
+			cliRes, err := ctx.client.CreateSession(ctx.ctxBase, cliPrm)
+			if err != nil {
+				return fmt.Errorf("default session: %w", err)
+			}
+
+			ctx.sessionToken = sessionTokenForOwner(owner.NewIDFromPublicKey(&ctx.key.PublicKey), cliRes)
+
+			err = ctx.sessionToken.Sign(ctx.key)
+			if err != nil {
+				return fmt.Errorf("sign token: %w", err)
+			}
+
+			_ = p.cache.Put(cacheKey, ctx.sessionToken)
+		}
+	}
+
+	return nil
+}
+
+func (p *pool) PutObject(ctx context.Context, hdr object.Object, payload io.Reader, opts ...CallOption) (*oid.ID, error) {
 	cfg := cfgFromOpts(append(opts, useDefaultSession())...)
-	cp, options, err := p.conn(ctx, cfg)
+
+	var ctxCall callContext
+
+	ctxCall.ctxBase = ctx
+
+	err := p.prepareCallContext(&ctxCall, cfg)
 	if err != nil {
 		return nil, err
 	}
 
-	res, err := cp.client.PutObject(ctx, params, options...)
+	var prm client.PrmObjectPutInit
 
-	if p.checkSessionTokenErr(err, cp.address) && !cfg.isRetry {
-		opts = append(opts, retry())
-		return p.PutObject(ctx, params, opts...)
+	wObj, err := ctxCall.client.ObjectPutInit(ctx, prm)
+	if err != nil {
+		return nil, fmt.Errorf("init writing on API client: %w", err)
 	}
 
+	wObj.UseKey(*ctxCall.key)
+
+	if ctxCall.sessionToken != nil {
+		wObj.WithinSession(*ctxCall.sessionToken)
+	}
+
+	if cfg.btoken != nil {
+		wObj.WithBearerToken(*cfg.btoken)
+	}
+
+	if wObj.WriteHeader(hdr) {
+		sz := hdr.PayloadSize()
+
+		if data := hdr.Payload(); len(data) > 0 {
+			if payload != nil {
+				payload = io.MultiReader(bytes.NewReader(data), payload)
+			} else {
+				payload = bytes.NewReader(data)
+				sz = uint64(len(data))
+			}
+		}
+
+		if payload != nil {
+			const defaultBufferSizePut = 4096 // configure?
```

Comment:
We used [3 MiB chunks](https://github.com/nspcc-dev/neofs-sdk-go/blob/b6abb02acec4d081f6a2c2abf0f845dd350fd749/client/object.go#L109) so they will fit default 4MiB gRPC message without split (gRPC message also includes bytes or protobuf format encoding). I propose to keep it the same.